compiler/rustc_middle/src/ty/sty.rs

```diff
@@ -1133,7 +1133,10 @@ impl<'tcx> ProjectionTy<'tcx> {
     /// then this function would return a `T: Iterator` trait reference.
     pub fn trait_ref(&self, tcx: TyCtxt<'tcx>) -> ty::TraitRef<'tcx> {
         let def_id = tcx.associated_item(self.item_def_id).container.id();
-        ty::TraitRef { def_id, substs: self.substs.truncate_to(tcx, tcx.generics_of(def_id)) }
+        // Include substitutions for generic arguments of associated types
+        let assoc_item = tcx.associated_item(self.item_def_id);
+        let substs_assoc_item = self.substs.truncate_to(tcx, tcx.generics_of(assoc_item.def_id));
+        ty::TraitRef { def_id, substs: substs_assoc_item }
```

Comment:
Ugh this has the same "problem" we discussed in `astconv/mod` (extra substs for the trait `DefId`). Where is this function used?

Reply:
I modified this for [this](https://github.com/rust-lang/rust/blob/e708cbd91c9cae4426d69270248362b423324556/compiler/rustc_typeck/src/astconv/mod.rs#L1269) call. But seems wrong.

Reply:
But I talked to Matthew about this and he told me we would need to keep the substs for the assoc type here.

Reply:
Ugh this weird. Can you just add a FIXME here? (something along the lines of this function probably should exist at all since it isn't clear what we really want).

The call that you linked works because we *only* look at the `substs`. But this is scary and easy to misuse.